junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestInstances.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.extension;
+
+import static org.apiguardian.api.API.Status.MAINTAINED;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.apiguardian.api.API;
+
+/**
+ * {@code ParameterContext} encapsulates the <em>test instances</em> of a test.
+ *
+ * <p>While top-level tests only have a single test instance, nested tests
+ * have one instance per containing test class.
+ *
+ * @since 5.4
+ * @see ExtensionContext#getTestInstances()
+ * @see ExtensionContext#getRequiredTestInstances()
+ */
+@API(status = MAINTAINED, since = "5.4")
+public interface TestInstances {
+
+	/**
+	 * Get the innermost test instance.
+	 *
+	 * <p>The innermost instance is the one closest to the test method.
+	 */
+	Object getInnermost();
+
+	/**
+	 * Get the enclosing test instances, excluding the innermost test instance.
+	 */
+	List<Object> getEnclosing();
+
+	/**
+	 * Get all test instances, ordered from outermost to innermost.
+	 */
+	List<Object> getAll();
+
+	/**
+	 * Find the test instance of the supplied required type.
+	 */
+	Optional<Object> find(Class<?> requiredType);
```

Comment:
Is this for finding test classes implementing a certain interface or extending a certain base class?

Reply:
According to the status quo in this proposal, it's an "instance of" check, so.... _both_. 😉

Reply:
If it is decided that this method should be kept, then I wonder if it could and should be changed to be something like the following, so that the test instance is already casted to the correct type when it's returned (if I've understood correctly what this method is meant to do):

```java
Optional<T> find(Class<T> requiredType);
```

@marcphillip Do you have any thoughts?

Reply:
@jbduncan Thanks for chiming in! That makes perfect sense and I changed it rightaway.

Reply:
@marcphillip Brilliant! Thank you for responding so quickly, and I'm glad that I understood enough to raise a worthwhile suggestion. :)